C client hangs when submitting concurrent requests

batiati opened the pull request:
- `Thread.available_messages` was being initialized with `message_pool.messages_max_client`.
This value accounts for more messages than is allowed by `config.client_request_queue_max`, causing the client to exceed the request queue.

- Added a `bool` return to `on_signal_fn`, indicating that there are still packets to be processed, causing the signal to be notified immediately.

Closes #230

batiati committed:
C_client available_messages fix

- Thread.available_messages was being initialized with message_pool.messages_max_client.
This value accounts for more messages than is allowed by config.client_request_queue_max, causing the client to exceed the request queue.

- Added a bool return to on_signal, indicating that there are still packets to be processed, causing the signal to be notified immediately.

sentientwaffle commented on src/c/tb_client/signal.zig:
```diff
@@ -274,8 +274,12 @@ pub const Signal = struct {
+
+            // Wakes up immediately if there are still packets to process.
+            if (retry) self.notify();
+            return;
```
I'm not at all familiar with `Signal` — if we are going to wake up immediately, why "wait" at all?

sentientwaffle replied:
Is it because we are trying due to having no available messages, so we must wait until there are available messages?

batiati replied:
It is supposed to run on the next tick, when ongoing requests may have finished, restoring the available messages.

Without that, the pending packets would be sent only if submit was called by the application.

sentientwaffle replied:
If no outgoing requests have finished by the next tick, how are they handled?

batiati replied:
It will just repeat again on the next tick. Maybe we should add an 'if' at the beginning of the function to return early without trying to iterate.

Lets rephrase this comment too, I think "immediately" isn't the best word.

batiati committed:
Review adjustments

- Avoid trying to process packets if there are no available messages.
- Better variable names and comments.

kprotty commented on src/c/tb_client/signal.zig:
```diff
@@ -274,8 +274,12 @@ pub const Signal = struct {
+            self.wait();
+
+            // Run again on the next tick if there are still packets to process.
+            if (pending) self.notify();
```
I think the `notify()` should be done before the `wait()` to go through the yield path of `io.timeout` (observing `.notified` at the CAS) instead of going through `io.recv` (which results in a syscall on non-io_uring platforms).

batiati replied:
Calling `notify()` before `wait()` hangs too.
It's not intuitive for me, but this time it printed this message after a few seconds running:

```
info(message_bus): message queue for peer .tigerbeetle.lib.tigerbeetle.src.message_bus.union:375:19{ .replica = 0 } full, dropping ping message
```

Strange enough, printing a debug message to stderr causes that same code to work most of the time, but it still hangs eventually.

```diff
+      std.log.err("timeout ...", .{});
        self.io.timeout(
            *Signal,
            self,
            on_timeout,
            &self.completion,
            0, // zero-timeout functions as a yield
        );
```

kprotty replied:
After some thought, should the notify() happen in `on_complete`? Thats where the available_messages is incremented again and is where a retry should occur

batiati replied:
Yes, I've thought of that, but it would be called multiple times needless, if there are many requests being completed on the same tick.

sentientwaffle replied:
> if there are many requests being completed on the same tick.

Since a client only has one request in-flight at a time, is that possible?

batiati replied:
Let me try to call notify only when available_messages == 1

kprotty replied:
notify() is cheap to call as it only does a syscall if the signal is waiting. If notify() is happning in the ClientThread code, the Signal is running the callback and not waiting.

batiati replied:
Much better now, thank you!

kprotty reviewed (commented):
Nice catch. There's a pending optimization that could be done, but otherwise this looks like it should correctly handle requeueing the signal callback.

batiati commented:
> Nice catch. There's a pending optimization that could be done, but otherwise this looks like it should correctly handle requeueing the signal callback.

@kprotty 
Probably I'm going to sound naive, but **why do we need a loopback socket for the `Signal` implementation?**

Since it runs on the same thread is ticking the IO, wouldn't it be safe to just iterate the `Stack` (which is already thread-safe) during the `run` function?

kprotty commented:
The loopback socket is used to wake up the tb_client OS thread from a language client's thread when the former sleeping in IO. Its equivalent to a crossplatform `eventfd` on linux or `pipe` on posix. This allows packets to be processed after the latest `notify()` instead of every `config.tick_ms` if `Stack` was drained in `run()`. In this case `IO.run_for_ns` is only used to facilitate blocking for IO (instead of busy-waiting with `tick()`) and consistently ticking the vsr.Client regularly even after / in the absence of `on_signal()`. 

TL;DR - it could be done in run() but may increase client latency. That might be fine though, would remove the need for Signal and make the code a lot simpler.

kprotty commented:
Removing Signal and relying on run_for_ns() is definitely worth exploring whether in this PR or in another. Especially when (apart from its simplicity), it's the method used in other clients like `demo.zig` and `benchmark.zig`.

batiati commented:
> Removing Signal and relying on run_for_ns() is definitely worth exploring whether in this PR or in another. Especially when (apart from its simplicity), it's the method used in other clients like `demo.zig` and `benchmark.zig`.

Great, let me try it comparing with the benchmark that I have implemented for the Java Client.

batiati committed:
Review adjustments

- Call notify() at the request completion, not at on_signal()

batiati commented:
> Removing Signal and relying on run_for_ns() is definitely worth exploring

@kprotty 
There are some benchmarks (using only 5 batches to avoid lsm compaction)

**benchmark.zig:**
```
Benchmarking...
============================================
546066 transfers per second
max p100 latency per 8191 transfers = 22ms
```

**Java client benchmark, using the status-quo tb_client:**
```
Benchmarking...
============================================
476220 transfers per second
create_transfers max p100 latency per 8191 transfers = 25ms
total 40955 transfers in 86ms
```

**Java client benchmark, tb_client modified without `Signal`**
```
Benchmarking...
============================================
319960 transfers per second
create_transfers max p100 latency per 8191 transfers = 33ms
total 40955 transfers in 128ms
```

that's the modification:

```zig
fn run(self: *Self) void {
    while (!self.signal.is_shutdown()) {
        self.client.tick();
        self.send_requests();
        self.io.run_for_ns(config.tick_ms * std.time.ns_per_ms) catch |err| {
            log.err("IO.run() failed with {}", .{err});
            return;
        };                
    }
}

fn send_requests(self: *Self) void {
    // Consume all of retry here to avoid infinite loop
    // if the code below pushes to self.retry while we're dequeueing.
    var pending = self.retry;
    self.retry = .{};

    // The loop below can exit early without processing all of pending
    // if available_messages becomes zero.
    // In such a case we need to restore self.retry we consumed above
    // with those that weren't processed.
    defer {
        pending.push(self.retry);
        self.retry = pending;
    }

    // Process packets from either pending or submitted as long as we have messages.
    while (self.available_messages > 0) {
        const packet = pending.pop() orelse self.submitted.pop() orelse break;
        const message = self.client.get_message();
        defer self.client.unref(message);

        self.available_messages -= 1;
        self.request(packet, message);
    }
}
```

kprotty commented:
Predicted latency getting hit so that seems to match up. What happens if you do `send_requests()` before `client.tick()`? The former may do `self.client.request()` which could queue up work for `client.tick()` to execute. I assume it wont have much of an effect but may increase throughput

sentientwaffle commented:
FYI you can increase `lsm_batch_multiple` (in `config.zig`) to give yourself more ops before compaction is triggered — I think you can set it as high as ½ of the `journal_slot_count` (so, `1024/2 = 512`). Be warned that setting it higher increases each replica's RAM requirements, though, so maybe try 32 or 64 first.

batiati commented:
> What happens if you do `send_requests()` before `client.tick()`?

Same result, it appears to slightly reduce the latency sometimes, but not consistently.

I guess I can include this change on this PR 😉

```diff
/// A Signal is a way to trigger a registered callback on a tigerbeetle IO instace 
/// when notification occurs from another thread.
/// It does this by using OS sockets (which are thread safe) 
/// to resolve IO.Completions on the tigerbeetle thread.
- ///
- /// TODO: implement a simpler version of this eventually..
pub const Signal = struct {

```

batiati committed:
Remove TODO comment
